test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc

```diff
@@ -0,0 +1,187 @@
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake-step-controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_THAT(status.message(), HasSubstr("Negative fixed_rps_value"));
+}
+
+TEST(FakeStepControllerConfigFactory, ValidateConfigWithDefaultConfigReturnsOk) {
+  FakeStepControllerConfig config;
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake-step-controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_TRUE(status.ok());
+}
+
+TEST(FakeStepControllerConfigFactory, ValidateConfigWithValidConfigReturnsOk) {
+  FakeStepControllerConfig config;
+  config.set_fixed_rps_value(1);
+  Envoy::ProtobufWkt::Any config_any;
+  config_any.PackFrom(config);
+  auto& config_factory =
+      Envoy::Config::Utility::getAndCheckFactoryByName<StepControllerConfigFactory>(
+          "nighthawk.fake-step-controller");
+  absl::Status status = config_factory.ValidateConfig(config_any);
+  EXPECT_TRUE(status.ok());
+}
+
+TEST(FakeStepController, GetCurrentCommandLineOptionsReturnsRpsFromConfig) {
+  FakeStepControllerConfig config;
+  config.set_fixed_rps_value(5678);
+  FakeStepController step_controller(config, CommandLineOptions());
+  EXPECT_EQ(step_controller.GetCurrentCommandLineOptions().value().requests_per_second().value(),
+            5678);
+}
+
+TEST(FakeStepController, IsConvergedInitiallyReturnsFalse) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsFalseAfterBenchmarkResultWithoutPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_FALSE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsConvergedReturnsTrueAfterBenchmarkResultWithPositiveScore) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  MetricEvaluation* evaluation = benchmark_result.mutable_metric_evaluations()->Add();
+  evaluation->set_threshold_score(1.0);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  EXPECT_TRUE(step_controller.IsConverged());
+}
+
+TEST(FakeStepController, IsDoomedReturnsFalseAfterSuccessfulBenchmarkResult) {
+  FakeStepController step_controller(FakeStepControllerConfig{}, CommandLineOptions{});
+  BenchmarkResult benchmark_result;
+  benchmark_result.mutable_status()->set_code(::grpc::OK);
+  step_controller.UpdateAndRecompute(benchmark_result);
+  std::string doomed_reason;
+  EXPECT_FALSE(step_controller.IsDoomed(doomed_reason));
```

Comment:
Reading this, Ibwonder if a single call to obtain status (converged) or failure (doomed) would simplify things here and for the consing code. Just a thought after recently learning about that construct, please feel free to ignore

Reply:
I agree with that. And even if you want to keep the IsDoomed call, you can still just have that refer to the status. Especially since Doomed and Converged are mutually exclusive states.

Reply:
Opened #466